### PR TITLE
Toplevel Parser Module

### PR DESCRIPTION
**Pull request by vb000**
A toplevel module to generate essential information with a single command:
```
python3 -m vanilla_parser
```

At the same time, provide full flexibility with all available command line flags:

```
python3 -m vanilla_parser <non-default log filenames> <additional flags>
```

### CODE REVIEW

**Comment by vb000** [comment]
I think this is ready and provides this interface:
```
usage: vanilla_parser [-h] [--trace TRACE] [--stats STATS] [--log LOG]
                      [--vcache-trace VCACHE_TRACE]
                      [--vcache-stats VCACHE_STATS] [--tile TILE]
                      [--tile-group TILE_GROUP] [--abstract] [--generate-key]
                      [--cycle CYCLE] [--only [SUBMODULE [SUBMODULE ...]]]
                      [--also [SUBMODULE [SUBMODULE ...]]] [--no-blood-graph]
                      [--per_vcache] [--no-stall-graph]

The interface for parsing COSIM output logs. This parser comprises
of following five tools:

- blood_graph
- stats_parser
- pc_histogram
- trace_parser
- vcache_stall_graph

By default blood_graph, stats_parser and pc_histogram are executed,
given required files are generated by cosim. If required files for
a tool are missing it is skipped.

optional arguments:
  -h, --help            show this help message and exit
  --trace TRACE         Vanilla operation log file
  --stats STATS         Vanilla stats log file
  --log LOG             Vanilla log file
  --vcache-trace VCACHE_TRACE
                        Vanilla operation log file
  --vcache-stats VCACHE_STATS
                        Vanilla stats log file
  --tile TILE           Also generate per tile stats
  --tile-group TILE_GROUP
                        Also generate per tile group stats
  --abstract            Type of graphs - abstract / detailed
  --generate-key        Generate a key image with graphs
  --cycle CYCLE         Cycle window of bloodgraph/stallgraph as
                        start_cycle@end_cycle.
  --only [SUBMODULE [SUBMODULE ...]]
                        List of tools to run instead of the default set
  --also [SUBMODULE [SUBMODULE ...]]
                        List of tools to run in addition to the default set

Blood graph specific options:
  --no-blood-graph      Skip blood graph generation

Stats parser specific options:
  --per_vcache          Also generate separate stats files for each victim
                        cache bank.

Vcache stall graph specific options:
  --no-stall-graph      Skip stall graph generation
```

By default we run `blood_graph`, `stats_parser` and `pc_histogram`. `vcache_stall_graph` and `trace_parser` can be included or exclusively run using `--only` and `--also` options.

File: software/py/vanilla_parser/__init__.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import sys,os
+sys.path.append(os.path.join(os.path.dirname(__file__)))
```
Comment (drichmond) [inline software/py/vanilla_parser/__init__.py:2]: Is there another way to do this? Can you do something like `from . import foo`?

**Comment by drichmond** [comment]
Let's get #308 merged, and double check that these work together (No reason why they shouldn't...)

Is your large comment above in a README, or somewhere?

File: software/py/vanilla_parser/__init__.py (hunk 1 above)
Comment (vb000) [reply software/py/vanilla_parser/__init__.py]: Apparently I can. Thanks Dustin, learnt a new thing today :)

**Comment by vb000** [comment]
> Is your large comment above in a README, or somewhere?

It's in the package's help message, would that do?

```
python3 -m vanilla_parser -h
```
prints the help message...

File: software/py/vanilla_parser/__init__.py (hunk 1 above)
Comment (drichmond) [reply software/py/vanilla_parser/__init__.py]: Glad I could help

**Comment by drichmond** [comment]
> > Is your large comment above in a README, or somewhere?
> 
> It's in the package's help message, would that do?
> 
> ```
> python3 -m vanilla_parser -h
> ```
> 
> prints the help message...

I feel like it's time to have a README for the parsers. Even if it's super simple, a mapping parser->.csv/.log and a message telling people to run with -h to get the command outputs.

You can either add it here, or file an issue. Your choice.

**Comment by vb000** [comment]
Oh mapping to log files, makes sense, it'd be very helpful to get started on these set of parsers.

**Comment by vb000** [comment]
Added a readme.

**Comment by drichmond** [comment]
Rebase, test and good to go

**Comment by vb000** [comment]
@bornaehsani Changes look good to you?

**Comment by bornaehsani** [comment]
Hey @vb000 , this all looks great to me! Amazing work! The interface looks great, you just made calling the profiler tools 10 times easier! The status messages it shows when calling each individual profiler tools is also very useful. Only comment I have is to have it rebased as we discussed, to include the recent changes made in PR #308, and then we're good to go! Oh and also please update the README for the PYTHONPATH, so it's clear that it shouldn't point to the vanilla_parser directory, but the parent directory.  Great work!

**Comment by vb000** [comment]
> Hey @vb000 , this all looks great to me! Amazing work! The interface looks great, you just made calling the profiler tools 10 times easier! The status messages it shows when calling each individual profiler tools is also very useful. Only comment I have is to have it rebased as we discussed, to include the recent changes made in PR #308, and then we're good to go! Oh and also please update the README for the PYTHONPATH, so it's clear that it shouldn't point to the vanilla_parser directory, but the parent directory. Great work!

Done.

**Pull request merged**